keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,462 @@
+        x-kubernetes-immutable: true
+      "y":
+        type: string
+        x-kubernetes-immutable: true
+```
+
+for all fields `x` and `y` of `foo`.
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+    properties:
+      bar:
+        type: object
+        x-kubernetes-immutable: true
+```
+
+can be simplified to 
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+    properties:
+      bar:
+        type: object
+```
+
+At the root-level and inside `.metadata` of an object, `x-kubernetes-immutable` is forbidden.
+
+### OpenAPI extension `x-kubernetes-immutable-keys`
+
+The topology of arrays is specified via `x-kubernetes-list-type` and `x-kubernetes-list-map-keys`. We currently support the `atomic`, `set` and `map` types. These interact with immutability in the following way:
+
+- **Non-recursive** immutability: For an array or a map, one can not remove or add new items/values, but existing items/values can be modified. 
+
+  - For maps this means that the the keys stay the same, but values might change:
+    ```yaml
+    properties:
+      someMap:
+        type: object
+        x-kubernetes-immutable-keys: true # allowed with additionalProperties
+        additionalProperties:
+          type: string
+    ```
+    with `x-kubernetes-immutable-keys` allowed due to `additionalProperties` and mutually
+    exclusive with `x-kubernetes-immutable` on the same node.
+    
+  - For arrays of list-type `map` we get the same behaviour, but enforce that the map-keys
+    are immutable as well (via CRD validation):
+    
+    ```yaml
+    properties:
+      someArray:
+        type: array
+        x-kubernetes-list-type: map
+        x-kubernetes-list-map-keys: ["name"]
+        x-kubernetes-immutable-keys: true # allowed with `x-kubernetes-list-type: map`
+        items:
+          properties:
+            name:
+              type: string
+              x-kubernetes-immutable: true # enforced
+            x:
+              type: string
+            y:
+              type: integer
+     ```
+     with `x-kubernetes-immutable-keys` allowed due to `x-kubernetes-list-type: map`.
+     
+- **Allowed addition and deletion**: New/deleted keys in maps are tolerated.
+
+  - For maps, this can be expressed by marking the values immutable,
+    but not the map:
+    ```yaml
+    properties:
+      someMap:
+        type: object
+        additionalProperties:
+          type: string
+          x-kubernetes-immutable: true
+    ```
+  - For arrays an identifier field is required, e.g. `.name`. We support this by reusing `x-kubernetes-list-type: map` and `x-kubernetes-list-map-keys` specifying the keys:
+    ```yaml
+    properties:
+      someArray:
+        type: array
+        x-kubernetes-list-type: map
+        x-kubernetes-list-map-keys: ["name"]
+        items:
+          properties:
+            name:
+              type: string
+              x-kubernetes-immutable: true
```

Comment:
I'm not sure I understand what this means

Reply:
We can move these up to the `items` level. Both are equivalent.

Reply:
actually, there is a slight difference between both:
```
type: object
nullable: true
properties:
  x:
    type: string
    x-kubernetes-immutable: true
```
allows a change from `{x: something}` to `null` or undefined (when used as a field value), but

```
type: object
x-kubernetes-immutable: true
nullable: true
properties:
  x:
    type: string
```

does not allow that.

Without `nullable` and as array items, both coincide.